feature(ddsp):

EdwinToh opened the pull request:
- initial commit of DDSP integration
- added methods for getAudioFeatures and toneTransfer
- added demo
- added models to repo for now, will move to checkpoints

notwaldorf commented:
FYI we should upload the models to a cloud storage bucket rather than checking them in the repo itself. I can help with that if needed

jesseengel commented:
The checkpoints are up at https://storage.googleapis.com/magentadata/js/checkpoints/ddsp
for example: https://storage.googleapis.com/magentadata/js/checkpoints/ddsp/flute/model.json

jesseengel commented on music/demos/ddsp.ts:
```diff
@@ -0,0 +1,176 @@
+import { MODEL } from '../src/ddsp/constants';
+import * as mm from '../src/index';
+
+function floatTo16BitPCM(output, offset, input) {
```
It seems like some of these functions would go good in a general "audio utils" library. What do you think @notwaldorf @adarob ?

EdwinToh replied:
this is a function used in the demo actually, I don't think the library should try to handle this but I'll leave it up to you guys to decide.

jesseengel commented on music/src/ddsp/audio_utils.ts:
```diff
@@ -0,0 +1,117 @@
+/**
```
Yah, there's nothing ddsp specific about these. It might be nice to centralize all the audio_utils in the repo if that makes sense.

jesseengel commented on music/src/ddsp/loudness_utils.ts:
```diff
@@ -0,0 +1,72 @@
+    .sqrt()
+    .squeeze();
+
+  // Amplitude to Db
```
It might be nice to split this out as a separate amplitude_to_db() function in an audio util library.

jesseengel commented on music/src/ddsp/loudness_utils.ts:
```diff
@@ -0,0 +1,72 @@
+    return [];
+  }
+
+  // Compute RMS energy
```
Same with this compute_rms_power() function

That would then help isolate this wrapper that does shifting and clipping on top of these two more fundamental functions.

jesseengel commented on music/src/ddsp/model.ts:
```diff
@@ -0,0 +1,110 @@
+    model: MODEL,
+    audioFeatures: AudioFeatures
+  ): Promise<Float32Array> {
+    return await toneTransfer(model, audioFeatures);
```
not to be picky about this, but technically toneTransfer() is the process of both extracting the audio features and then synthesizing.

You could rename toneTransfer to synthesize, and then have another convenience function toneTransfer that just takes in audio and returns the tone transferred audio. 

Also, I'm not sure it makes sense to have the synthesize function (toneTransfer() at the moment) as a separate method, because it always needs to be paired with a model, so it might make sense to just have it defined in the class itself.

EdwinToh replied:
I thought a little bit about this and decided to do it this way because of "getAudioFeatures" is actually agnostic of the models, and if you see the demo, you can see that you just need to run getAudioFeatures once, and then run synthesize with all 4 different models. WDYT?

jesseengel replied:
That's fine, but then the method should be called `synthesize` because that's what it does :). I'm just saying if someone makes a demo with only one model and wants a convenience function that goes straight audio2audio then a convenience function named toneTransfer would make sense.

jesseengel commented on music/src/ddsp/pitch_utils.ts:
```diff
@@ -0,0 +1,191 @@
+import { Tensor } from '@tensorflow/tfjs';
+
+function upsample_linear(
+  spicePitchesOutput: number[],
```
you could name these a bit more generally, because it's actually just a linear interpolation function right?

jesseengel commented on music/src/ddsp/pitch_utils.ts:
```diff
@@ -0,0 +1,191 @@
+  return fmin * Math.pow(2.0, (1.0 * cqtBin) / binsPerOctave);
+}
+
+async function getPitches(spiceModel: tf.GraphModel, inputData: AudioData) {
```
since this is so bound to the specifics of the spiceModel, it might make sense for us to make our own SpiceModel class that has this function and all these SPICE constants as attributes.

jesseengel commented on music/src/ddsp/spice.ts:
```diff
@@ -0,0 +1,31 @@
+import * as tf from '@tensorflow/tfjs';
+import { SPICE_MODEL_URL } from './constants';
+
+async function startSpice() {
```
Yah, if you follow the same general model class API that the rest of magenta.js uses for the spice model as well, I think it would be clearer how to use it and more modular.

jesseengel commented on music/src/ddsp/model.ts:
```diff
@@ -0,0 +1,110 @@
+   *
+   * @param model The model to use
+   */
+  constructor() {
```
If we're following the model API for the rest of magenta.js I think it makes sense to provide the constructor with the url and constants, for example: https://github.com/magenta/magenta-js/blob/master/music/src/gansynth/model.ts#L30

jesseengel reviewed (commented):
Looks pretty good Edwin, awesome job!

I just left a couple comments on how I think restructuring to follow the common api in magenta.js for models (for both DDSP and SPICE) could help make things clearer and more modular.

Also, (less pressing) as a broader comment, it would definitely be nice to incorporate all the non ddsp specific audio_util functions into https://github.com/magenta/magenta-js/blob/master/music/src/core/audio_utils.ts (but that goes for other models too, so that could be a refactor later).

notwaldorf commented on music/package.json:
```diff
@@ -8,14 +8,15 @@
   "unpkg": "dist/magentamusic.js",
   "dependencies": {
-    "@tensorflow/tfjs": "^1.1.2",
+    "@tensorflow/tfjs": "^2.7.0",
```
have you double checked everything else works when we switch to 2.x? Like, are all the tests passing, etc?

EdwinToh replied:
Not yet, but definitely something I am planning to do.

notwaldorf commented on music/report.20201105.152937.93397.0.001.json:
```diff
@@ -0,0 +1,689 @@
+
```
I don't know what this file is but it probably shouldn't be checked in

notwaldorf commented on music/src/ddsp/add_reverb.ts:
```diff
@@ -0,0 +1,59 @@
+import * as Tone from 'tone';
```
needs a license

notwaldorf commented on music/src/ddsp/buffer_utils.ts:
```diff
@@ -0,0 +1,31 @@
+export const arrayBufferToAudioBuffer = (
```
needs a license

notwaldorf commented on music/src/ddsp/constants.ts:
```diff
@@ -0,0 +1,68 @@
+export const MIN_VRAM = 50; // used for memory check
```
needs a licence

notwaldorf commented on music/src/ddsp/interfaces.d.ts:
```diff
@@ -0,0 +1,18 @@
+export interface AudioFeatures {
```
needs a licence

notwaldorf commented on music/src/ddsp/model.ts:
```diff
@@ -0,0 +1,110 @@
+   * Disposes of any untracked `Tensors` to avoid GPU memory leaks.
+   */
+  dispose() {
+    if (!this.initialized) {
```
we really don't have to dispose of anything here?

notwaldorf commented on music/src/ddsp/pitch_utils.ts:
```diff
@@ -0,0 +1,191 @@
+  }
+
+  // @ts-ignore
+  output[0].dispose();
```
Why are there so many ts-ignores? I don't think normally we need to ignore the tensor disposing

EdwinToh replied:
@notwaldorf I have a TODO to fix types (which I privately DM'ed Jesse) so apologies for not stating it in this draft PR! (It's also why this is still a draft PR and not ready for a review)

In this case, I'm actually struggling a little with TFJS and TS where they expect a Tensor in the return of a prediction but our models actually return an array with 2 tensors.

notwaldorf commented on music/src/ddsp/add_reverb.ts:
```diff
@@ -0,0 +1,59 @@
+  const bufferWithReverbData = bufferWithReverb.getChannelData(0);
+  // manual cleanup for safari
+  resampledAudioBuffer = null;
+  // @ts-ignore
```
Hmm, why are there so many ignores?

EdwinToh replied:
@notwaldorf I have a TODO to fix types (which I privately DM'ed Jesse) so apologies for not stating it in this draft PR!  (It's also why this is still a draft PR and not ready for a review)

In this case, I'm actually struggling a little with TFJS and TS where they expect a Tensor in the return of a prediction but our models actually return an array with 2 tensors.

notwaldorf commented on music/src/ddsp/audio_utils.ts:
```diff
@@ -0,0 +1,117 @@
+
+function shiftF0(f0Hz: number[], f0OctaveShift = 0.0) {
+  return tf.tidy(
+    // @ts-ignore
```
also here

notwaldorf reviewed (changes requested):
Some initial cleanup comments, I haven't actually looked at any of the code yet

EdwinToh commented:
@jesseengel @notwaldorf This PR is now ready for review. I've addressed all of the comments, except for moving some audio functions into core.

While I wait for your review, I will be testing (and fixing) the other demoes/libraries to make sure that the upgrade to TFJS and Tone.js will not break them.

EdwinToh commented:
i've ran tests and manually tested all the demoes, everything is working. Just FYI, there're a few pre-existing type errors on Music Vae, Core, and Piano Genie that I did not fix since I didn't want this PR to touch those files.

jesseengel commented on music/demos/ddsp_tone-transfer.html:
```diff
@@ -0,0 +1,92 @@
+<!-- Copyright 2020 Google Inc. All Rights Reserved.
```
super minor nit, but can you keep all snake case for the filename?

jesseengel commented on music/demos/ddsp_tone-transfer.ts:
```diff
@@ -0,0 +1,209 @@
+/**
```
same here

jesseengel commented on music/demos/ddsp_tone-transfer.ts:
```diff
@@ -0,0 +1,209 @@
+export const MODEL_URL =
+  'https://storage.googleapis.com/magentadata/js/checkpoints/ddsp';
+
+export const PRESET_MODELS = {
```
Another super minor nit, but these are PRESET_MODEL_URLS technically right?

jesseengel commented on music/demos/index.html:
```diff
@@ -55,6 +55,7 @@ <h4>
         <li><a href="piano_genie_conditioned.html">Play Piano Genie with specified key signatures and/or chord progressions</a></li>
         <li><a href="gansynth.html">Neural audio synthesis with GANSynth</a></li>
         <li><a href="coconet.html">Infill Bach-style 4-part harmonies with Coconet</a></li>
+        <li><a href="ddsp_tone-transfer.html">Transform audio into other musical instruments using DDSP</a></li>
```
same here

jesseengel commented on music/src/ddsp/model.ts:
```diff
@@ -0,0 +1,106 @@
+    audioFeatures: AudioFeatures,
+    options?: ModelValues
+  ): Promise<Float32Array> {
+    let modelValues = getModelValues(this.checkpointUrl);
```
This doesn't make much sense to me, it makes the model a leaky abstraction (relying on the values in CONSTANTS). 

It seems like what we would want is to set all of these values in the constructor.

EdwinToh replied:
by putting the values here instead of the constructor, we give the users the ability to tweak the values without creating a new instance.

jesseengel replied:
Understood, I think the better way to do that is to have the method take in optional arguments on call, that use the model attributes as defaults if not provided.

jesseengel commented on music/src/ddsp/ddsp.ts:
```diff
@@ -0,0 +1,387 @@
+  return presetModelValues.length > 0 ? presetModelValues[0] : {};
+}
+
+async function synthesize(
```
Since this function only is used in the context of a model, wouldn't it make sense to just have it as a method of the Model class, rather than a separate function?

EdwinToh replied:
this is a method of the model class, or maybe i am missing your point. the only thing i did here is to separate it out into it's own file to keep the model class a little bit more lightweight. i can move it all into the same file.

jesseengel replied:
yah, that's it. It just makes sense to have all the code together, because these are all the things specific to the model. (i.e. `model` is always `self`)

jesseengel commented on music/src/ddsp/ddsp.ts:
```diff
@@ -0,0 +1,387 @@
+  });
+}
+
+async function ddsp_process(
```
Same for this one, it only really makes sense in the context of a model so it seems cleaner to just make it a method of the model class.

jesseengel commented on music/src/ddsp/ddsp.ts:
```diff
@@ -0,0 +1,387 @@
+  return { f0_hz, loudness_db };
+}
+
+export function getModelValues(checkpointUrl: string): ModelValues {
```
This is part of that leaky abstraction, and should probably take place in the model class constructor.

jesseengel commented on music/src/ddsp/constants.ts:
```diff
@@ -0,0 +1,77 @@
+export const CROSSFADE_DURATION = 1;
+export const OUTPUT_SAMPLE_RATE = 48000;
+
+export const PRESET_MODELS = [
```
More about the leaky abstraction / constructor. Taking a new look at this, I don't think having these as constants in the javascript code makes a lot of sense.

We'd like the ability to let people train their own models soon, and this would require them to add these values to the magenta-js code which doesn't work well for dropping in new models at runtime. 

I think it might make sense to have these in a json object (stored on cloud) that is read as well as the model weights. That way we can have our training scripts create these extra json files as well, and then users can just point to the URL of the directory with all the model info in it. 

I think ideally, the model constructor would take the URL and these values in it's constructor, but then override these values with values from the JSON if it's found.

We could do this as a later PR if you want, but I think it's important to do, and would keep the Model class from being leaky.

EdwinToh replied:
i see what you mean now with the constructor, perhaps we can have it in both places. these options will be read in the constructor and can be overwritten in the synthesize method.

i agree with taking it out of the constants file and moving them somewhere else.   

now that i take a look at this again, i realize that there are no default values if the checkpointUrl doesn't match any of these preset models. perhaps each one of these values need to be required parameters in the constructor and we can list the preset models' values in documentation.

jesseengel replied:
yah, I think it makes sense to override with values read them from jsons (that we can create for new datasets and new models as well).

i.e. Order of overwriting:
====================
--> Constructor Defaults
--> Constructor Args
---> Values loaded from json read in constructor

EdwinToh replied:
I've moved these values out of the constants and created a new README to provide these values since we don't have a way to load them from a json alongside the model for now.

jesseengel reviewed (commented):
Getting there! Some more comments.

EdwinToh commented:
@jesseengel I've moved the constants out and added a README for DDSP where I provide the values for the 4 respective models. Let me know what you think?

jesseengel commented on music/src/ddsp/model.ts:
```diff
@@ -0,0 +1,243 @@
+        loudness_db: loudnessTensor,
+      });
+
+      // apply a loudness modifier here to make some
```
wouldn't it make sense to do this once on the stitched audio instead of on each chunk?

EdwinToh replied:
yes, good point! however, looking at the code, it seems like we're doing this mul on a tensor and then disposing it once we get the data. 

if we do it once at the end, we'll have to convert audioChunks into a tensor, run the mul, then take the data again.  not sure if there is a performance gain if we do that.

jesseengel replied:
okay, I was just looking for code clarity mostly. If you think it will look/be worse, no worries.

EdwinToh replied:
word. i've made the change. I realized that it was a simple multiplier and doesnt need a tensor to do this operation.

jesseengel commented on music/src/spice/spice.ts:
```diff
@@ -0,0 +1,77 @@
+import { getPitches } from './pitch_utils';
+
+// SPICE SPECIFIC CONSTANTS
+const SPICE_MODEL_URL = 'https://tfhub.dev/google/tfjs-model/spice/2/default/1';
```
I like that you made a spice model class. Wouldn't it make sense to have these be default args of the Spice model, rather than in this separate file? and to have the functions be methods of the model? That way it has the same api as the other models (takes in a url in the constructor, but with a default to this url).

EdwinToh replied:
I think it makes more sense to have this here since this constant won't be modified (and shouldn't be). I think to try to make this class be consistent with the other classes in the Magenta library is a bad idea in this case because the url parameter will be optional and might potentially mislead one into thinking that they can use their own model somehow. 

I'm in favor to keep it this way unless you feel strongly about this.

EdwinToh replied:
on second thought, users might want to avoid using tfhub altogether. i've made the changes as you suggested!

jesseengel commented on music/src/ddsp/README.md:
````diff
@@ -0,0 +1,85 @@
+To start synthesizing, you'll need to pass in AudioFeatures (derived from the [SPICE](../spice) class).
+
+```
+violinDDSP.synthesize(audioFeatures);
````
can you also give an example of passing in a different say postGain or meanPitch?

jesseengel commented on music/src/ddsp/README.md:
````diff
@@ -0,0 +1,85 @@
+Each of these models need some settings to be passed. You can tweak these settings, but these are the defaults:
+
+```
+  // violin
````
Okay I've added all of these settings as `settings.json` files in each corresponding folder. I like that you provide documentation here on how to build with your own settings in the constructor, but I think the documentation and code should default to getting the URL of the folder (like in musicvae: https://github.com/magenta/magenta-js/blob/master/music/demos/music_vae.ts#L32), and then reading $URL/model.json and $URL/settings.json from the folder (sort of like in musicvae: https://github.com/magenta/magenta-js/blob/master/music/src/music_vae/model.ts#L847). The model should only error out if it didn't receive a settings in either the constructor or find and parse a settings.json in folder.

EdwinToh replied:
Nice! I've modified the class to do this now.

jesseengel commented on music/src/ddsp/README.md:
````diff
@@ -0,0 +1,85 @@
+To start using DDSP, create a new instance of DDSP and initialize it:
+
+```
+const violinDDSP = new DDSP(
````
It's cool to show an example of manually setting it, but it should be the second example, where the first is just

```
const violinDDSP = new DDSP(
  "https://storage.googleapis.com/magentadata/js/checkpoints/ddsp/violin"
);
await violinDDSP.initialize();
```

jesseengel reviewed (commented):
Okay, getting real close! I added `settings.json` files to cloud and left comments

EdwinToh commented:
@jesseengel @notwaldorf Looks like everything is addressed. Let me know if there are anymore things to take a look at!

notwaldorf commented on music/src/core/audio_utils.ts:
```diff
@@ -453,7 +504,7 @@ function outerSubtract(arr: Float32Array, arr2: Float32Array): Float32Array[] {
 
 function pow(arr: Float32Array, power: number) {
-  return arr.map(v => Math.pow(v, power));
+  return arr.map((v) => Math.pow(v, power));
```
What code formatter are you using? Is it the project default one (clang-format)? there's a lot of these meaningless formatting changes that I'm not sure where they're coming from...

EdwinToh replied:
I've re-saved all the files I touched with the clang-formatter and if you still see a lot of changes, it's because of the tonejs and tfjs upgrade. 

The change that touched the most file was the deprecation of `toMaster`, which required me to change it to `toDestination`.

notwaldorf replied:
Thanks! Looks good now!

notwaldorf commented on music/.gitignore:
```diff
@@ -5,3 +5,5 @@ es6/
 dist/
 yarn-error.log
+report.*.json
+tsconfig.json
```
why is the tsconfig being ignored? seems bad

EdwinToh replied:
afaik, vscode only takes tsconfig.json as the config but since this code repository only has the es5 and 6 flavors, i didn't want to include another tsconfig.json for myself just to see errors inline in my editor.

https://code.visualstudio.com/docs/typescript/typescript-compiling#_tsconfigjson

EdwinToh replied:
prob shd also link to the issue: https://github.com/Microsoft/vscode/issues/12463

notwaldorf commented on music/src/core/logging.ts:
```diff
@@ -15,23 +15,34 @@
-  WARN = 5,   // WARN messages will be logged.
-  INFO = 10,  // INFO and WARN messages will be logged.
-  DEBUG = 20  // DEBUG, INFO, and WARN messages will be logged.
+  NONE = 0, // No messages will be logged.
```
This code formatting change seems strictly worse, for example

EdwinToh replied:
i do have prettier setup and personally have always preferred the consistent formatting that prettier gives me and my team. i wasn't aware that the repo itself has another formatter since it never kicked in for me even when i disabled prettier. let me see whether i can setup clang and resave these files.

notwaldorf commented on music/src/core/player.ts:
```diff
@@ -409,29 +399,35 @@ class DrumKit {
+  }).toDestination();
 
-  private polySynth = new Tone.PolySynth(10).toMaster();
+  private polySynth = new Tone.PolySynth().toDestination();
```
why is the (10) argument not needed anymore?

EdwinToh replied:
it's a new tonejs version change: `PolySynth does not require a polyphony value. Voice allocation and disposing is done automatically based on demand.`

notwaldorf replied:
Got it, thanks!

notwaldorf commented on music/src/core/player.ts:
```diff
@@ -409,29 +399,35 @@ class DrumKit {
       this.drumKit.playNote(note.pitch, time, velocity);
     } else {
-      const freq = new Tone.Frequency(note.pitch, 'midi');
+      const freq = Tone.Frequency(note.pitch, 'midi').toFrequency();
```
Why is the 'new' not needed anymore?

EdwinToh replied:
because this is now a method and not a class

notwaldorf commented on music/src/core/player.ts:
```diff
@@ -522,9 +522,12 @@ export class SoundFontPlayer extends BasePlayer {
-           j++) {
-        ns.notes.push({pitch: i, velocity: j, program, isDrum});
+      for (
+        let j = constants.MIN_MIDI_VELOCITY;
```
This is also a pretty dubious formatting change

notwaldorf commented on music/webpack/es5.base.config.ts:
```diff
@@ -6,7 +6,7 @@ export const baseConfig = {
       test: /\.ts$/,
       exclude: /node_modules/,
-      use: {loader: 'ts-loader', options: {configFile: 'tsconfig.es5.json'}}
+      use: {loader: 'ts-loader', options: {configFile: 'tsconfig.es6.json'}}
```
I don't think this is correct. This is the es5 config file, and should be using the es5 rules, no?

EdwinToh replied:
upgrading tonejs (which now is in typescript) started to give errors and some googling gave me this result:
https://stackoverflow.com/questions/50203369/class-constructor-cannot-be-invoked-without-new-typescript-with-commonjs/50203532#50203532 specifically this: ```This problem should be solved in Node.js by setting TypeScript target option to es6 or higher. Modern Node.js versions support ES6 classes, there is no need to transpile them.
```

Happy to change this back to a working es5 tsconfig but I'll require some help from you to get it right.

notwaldorf replied:
Hmm, i don't think the node build should be using the es5.base.config. I am worried that if we're changing it here, then our es5 build will not be an es5 build anymore.

EdwinToh replied:
thanks for catching this. 

i changed this back to the es5 config and for some reason, I'm no longer running into the issues that I was having before! it's very perplexing, so I went through all the demoes to make sure everything still works and... they do. perhaps the issue I was having had to do with something else which I fixed.

notwaldorf reviewed (changes requested):
I've only really reviewed the non-ddsp changes as a library owner. It's super hard to follow this PR because it touches 37 files, a lot of them which only have code formatting changes, and most that don't seem to come from the google formatter (the default one). Do you maybe have prettier set up or something?

Please let me know if you can't get clang-format to work (should work if you open the magenta-js/music folder in vs code...)

notwaldorf commented on music/src/spice/model.ts:
```diff
@@ -0,0 +1,89 @@
+   * Loads variables from the checkpoint and builds the model graph.
+   */
+  async initialize() {
+    this.spiceModel = await startSpice(this.modelUrl);
```
shouldn't this set this.initialized?

EdwinToh replied:
good catch!

notwaldorf commented on music/src/ddsp/model.ts:
```diff
@@ -0,0 +1,263 @@
+      ...this.settings,
+    };
+    console.log('settings', settings);
+    console.log('this.settings', this.settings);
```
shouldn't this function set this.initialized = true?

EdwinToh replied:
good catch!

notwaldorf reviewed (changes requested):
more comments

jesseengel commented on music/checkpoints/checkpoints.json:
```diff
@@ -215,5 +215,33 @@
+    "id": "ddsp_flute",
+    "model": "DDSPFlute",
+    "description": "A flute model for use with DDSP.",
+    "url": "https://storage.googleapis.com/magentadata/js/checkpoints/ddsp/flute/model.json",
```
I think these URLs should be just to the folder right? Not to the model.json?

jesseengel commented on music/demos/ddsp_tone_transfer.ts:
```diff
@@ -0,0 +1,253 @@
+
+export const PRESET_MODELS = {
+  [MODEL.VIOLIN]: {
+    checkpointUrl: `${PRESET_MODEL_URL}/${MODEL.VIOLIN}`,
```
If you have the URL, you shouldn't need these settings, because you can just read them from settings.json off of cloud in the model directory, right?

jesseengel reviewed (commented):
Cool, looking pretty good! Just a couple small comments on my side

EdwinToh commented:
@jesseengel good catches. I've addressed your last 2 comments. 

@notwaldorf Thanks for your reviews so far. it'll be great if you can take another look again when you're free!

notwaldorf commented on music/demos/ddsp_tone_transfer.ts:
```diff
@@ -0,0 +1,201 @@
+export const PRESET_MODEL_URL =
+    'https://storage.googleapis.com/magentadata/js/checkpoints/ddsp';
+
+export const PRESET_MODELS = {
```
I don't think you really need this separate array. Since you use it as `PRESET_MODELS[MODEL.VIOLIN]`, that value is exactly just `${PRESET_MODEL_URL}/${MODEL.TRUMPET}`. Like, you're not getting anything from the array itself.

EdwinToh replied:
good point. this array had some settings before, but you're right that it doesn't make sense to keep it without those settings now.

notwaldorf commented on music/src/core/player.ts:
```diff
@@ -475,9 +480,10 @@ export class SoundFontPlayer extends BasePlayer {
+      programOutputs?: Map<number, any>,  // tslint:disable-line:no-any
+      drumOutputs?: Map<number, any>,     // tslint:disable-line:no-any
+      callbackObject?: BasePlayerCallback) {
+    // tslint:disable-line:no-any
```
i don't think this last comment is needed?

EdwinToh replied:
oh weird. this is clang doing some weird formatting but yeah it wasn't needed in the first place

notwaldorf commented on music/src/ddsp/constants.ts:
```diff
@@ -0,0 +1,28 @@
+export const CROSSFADE_DURATION = 1;
+export const OUTPUT_SAMPLE_RATE = 48000;
+
+// optimal values
```
nit: Capitalize comment, add period at the end

notwaldorf commented on music/src/ddsp/ddsp.ts:
```diff
@@ -0,0 +1,225 @@
+  const screenSize = window.screen.availWidth * window.screen.availHeight;
+  const DPI = window.devicePixelRatio;
+
+  // checking of vram ensures that the device can hold
```
nit: Capitalize comment, add period at the end (in whole file)

notwaldorf commented on music/src/ddsp/model.ts:
```diff
@@ -0,0 +1,232 @@
+      loudnessTensor.dispose();
+    }
+
+    // does not require stiching
```
nit: Capitalize comment, add period at the end (and rest of file)

notwaldorf commented on music/src/spice/index.ts:
```diff
@@ -0,0 +1,18 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc. All Rights Reserved.
```
2020

notwaldorf commented on music/src/spice/pitch_utils.ts:
```diff
@@ -0,0 +1,182 @@
+    }
+  }
+
+  // cover missing pitches
```
nit: Capitalize comment, add period at the end

notwaldorf reviewed (commented):
Looks great, thanks for your patience! I basically only left nits. I have a question: have you tested this on mobile? We've had a bunch of problems with tfjs on iphones in particular (float32 errors), and I want to make sure we test this while it's still fresh in someone's head

EdwinToh commented:
@notwaldorf Yep, we checked extensively when we released the Tone Transfer showcase. This works on mobile devices. The float32 issue has to do with iPhones only doing float16 I believe, and we've built against that.

notwaldorf reviewed (approved):
It's a :shipit: from me!!!

EdwinToh commented:
Awesome! Thanks @notwaldorf !

@jesseengel Will either one of you merge this PR? I don't have write access. :)
